eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,187 @@
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        let arrayElement = inputArray[currentIndex];
+        currentIndex--;
+        return innerArrayToList(inputArray, { value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(inputArray);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        let element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest:listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    let arrayAccumulator = [];
+
+    function innerListToArray(list) {
+        arrayAccumulator.push(list.value);
+        if (list.rest) {
+            innerListToArray(list.rest)
+        }
+        return arrayAccumulator;
+    }
+
+    return innerListToArray(inputList);
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArrayIterative(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+
+    let listLength = getListLength(inputList);
+    
+    let acc = [];
+    let listBuffer = inputList;
+    for (let i = 0; i < listLength; i++) {
+        if(listBuffer.value !== null) {
+            acc.push(listBuffer.value);
+        }
+        listBuffer = listBuffer.rest;
+    }
+
+    return acc;
+}
+
+/**
+ * Returns list from input list with prepended element
+ * @param {*} inputList 
+ * @param {*} element 
+ * @returns 
+ */
+function prepend(inputList, element) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    return {
+        value: element,
+        rest: inputList
+    }
+}
+
+/**
+ * Returns nth element in list
+ * @param {*} inputList 
+ * @param {*} targetIndex 
+ * @returns 
+ */
+function nth(inputList, targetIndex) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    if (isNaN(targetIndex) || targetIndex < 0 || Math.abs(targetIndex) === Infinity) {
+        throw "invalid target index";
+    }
+
+    listBuffer = inputList;
+    for (let i = 0; i < targetIndex; i++) {
+        listBuffer = listBuffer.rest;
+    }
+
+    if (listBuffer === null) {
+        throw "targetIndex out or range";
+    }
+
+    return listBuffer.value;
+}
+
+/**
+ * Returns true if input data structure is list
+ * @param {*} inputList 
+ * @returns 
+ */
+function isList(inputList) {
+    if (!Object.keys(inputList).includes('value') || !Object.keys(inputList).includes('rest') || Object.keys(inputList).length !== 2) {
```

Comment:
I think this would be a bit clearer as a function if you did a bit of boolean logic simplification. You could make it:
`return (keys includes value) && (keys includes rest) && (keys.length === 2)`